pkg/reloader/reloader.go

```diff
@@ -27,89 +28,153 @@ import (
 // It optionally substitutes environment variables in the configuration.
 // Referenced environment variables must be of the form `$(var)` (not `$var` or `${var}`).
 type Reloader struct {
-	logger           log.Logger
-	reloadURL        *url.URL
-	cfgFilename      string
-	cfgSubstFilename string
-	ruleDir          string
-	lastCfgHash      []byte
-	lastRulesHash    []byte
+	logger          log.Logger
+	reloadURL       *url.URL
+	cfgFile         string
+	cfgEnvsubstFile string
+	ruleDir         string
+	ruleInterval    time.Duration
+	retryInteval    time.Duration
+
+	lastCfgHash  []byte
+	lastRuleHash []byte
 }
 
 // New creates a new reloader that watches the given config file and rule directory
 // and triggers a Prometheus reload upon changes.
 // If cfgEnvsubstFile is not empty, environment variables in the config file will be
-// substituted and the out put written into the given path. Promethes should then
+// substituted and the out put written into the given path. Prometheus should then
 // use cfgEnvsubstFile as its config file path.
-func New(logger log.Logger, reloadURL *url.URL, cfgFile, cfgEnvsubstFile, ruleDir string) *Reloader {
+func New(logger log.Logger, reloadURL *url.URL, cfgFile string, cfgEnvsubstFile string, ruleDir string) *Reloader {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
 	return &Reloader{
-		logger:           logger,
-		reloadURL:        reloadURL,
-		cfgFilename:      cfgFile,
-		cfgSubstFilename: cfgEnvsubstFile,
+		logger:          logger,
+		reloadURL:       reloadURL,
+		cfgFile:         cfgFile,
+		cfgEnvsubstFile: cfgEnvsubstFile,
+		ruleDir:         ruleDir,
+		ruleInterval:    3 * time.Minute,
+		retryInteval:    5 * time.Second,
 	}
 }
 
-// Watch starts to watch the config file and processes it until the context
-// gets canceled.
+// Watch starts to watch the config file and rules and process them until the context
+// gets canceled. Config file gets env expanded if cfgEnvsubstFile is specified and reload is trigger if
+// config or rules changed.
 func (r *Reloader) Watch(ctx context.Context) error {
-	w, err := fsnotify.NewWatcher()
+	configWatcher, err := fsnotify.NewWatcher()
 	if err != nil {
 		return errors.Wrap(err, "create watcher")
 	}
-	defer w.Close()
-
-	cfgDir := filepath.Dir(r.cfgFilename)
+	defer configWatcher.Close()
 
-	if err := w.Add(cfgDir); err != nil {
-		return errors.Wrap(err, "add config file directory watch")
-	}
-	if _, err := r.applyConfig(); err != nil {
-		return errors.Wrap(err, "initial apply")
-	}
-	if err := r.triggerReload(ctx); err != nil {
-		level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+	initial := make(chan struct{}, 1)
+	if r.cfgFile != "" {
+		if err := configWatcher.Add(r.cfgFile); err != nil {
+			return errors.Wrap(err, "add config file watch")
+		}
+		level.Info(r.logger).Log(
+			"msg", "started watching config file for changes",
+			"in", r.cfgFile,
+			"out", r.cfgEnvsubstFile)
+		initial <- struct{}{}
```

Comment:
Mh, I'd prefer if we went back to not having all the logic in the for loop body but some methods instead. Then we can just call the function here once and then in the loop. Seems easier to grasp that using this `initial` channel here to trigger a first pass of all the code in the loop.

Reply:
mhm, sure I guess we can do so.